### PR TITLE
plot geometry for dispersive materials without initializing structure object

### LINKED ISSUE #1824: plot geometry directly (without discretization)
Would be nice to be able to plot the geometry directly, without creating the `structure` object and the discretization first (which can be slow for large simulations).

It's pretty tricky to implement in its full generality for all possible objects, but would be easy to support a subset of objects (e.g. prisms and other objects aligned with the grid).

The `get_epsilon_grid` function actually implements one possible version of this functionality, and we could simply plot it.

cc @joamatab

### PR DESCRIPTION
**Pull request by oskooi**
Closes #1824.

Removed the `frequency` parameter from `plot2D` since this cannot be passed to `get_epsilon_grid`. As a result, this API change is not backwards compatible. As a workaround, we can leave the `frequency` parameter in but just mention in the documentation that it is simply ignored.

cc @joamatab

### CODE REVIEW

**Comment by oskooi** [comment]
For the record, there was a single test failure unrelated to this PR that went away after rerunning. The symmetry unit test of `test_material_grid.py` was failing a check in the sixth decimal digit for single precision:
```
======================================================================
FAIL: test_symmetry (__main__.TestMaterialGrid)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/meep/meep/build/meep-1.22.0-beta/_build/sub/python/../../../python/tests/test_material_grid.py", line 161, in test_symmetry
    self.assertAlmostEqual(tran_nosym, tran_sym)
AssertionError: 20.27522531963953 != 20.275228328448193 within 7 places (3.0088086617752197e-06 difference)
```
It might therefore be useful to increase the tolerance slightly for this test to avoid this failure in the future.

**Comment by oskooi** [comment]
To demonstrate that this feature is working correctly, here is a 2d test case which contains a variety of different simulation objects: two types of geometric objects (`Cylinder` and `MaterialGrid`), a source, PML in a single direction, and a DFT monitor. `plot2D` is called with three different `resolution` values of 20, 80, and 320 for sampling the ε grid. A single-threaded run takes about ~0.5 s because the `structure` object is never initialized.

ring_matgrid_res20

ring_matgrid_res80

ring_matgrid_res320

```py
import meep as mp
import numpy as np
import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as plt

mp.verbosity(3)

cell_size = mp.Vector3(1,1,0)

design_region_resolution = 500
design_shape = mp.Vector3(1.0,1.0,0)
Nx = int(design_region_resolution*design_shape.x) + 1
Ny = int(design_region_resolution*design_shape.y) + 1
x = np.linspace(-0.5*cell_size.x,0.5*cell_size.x,Nx)
y = np.linspace(-0.5*cell_size.y,0.5*cell_size.y,Ny)
xv, yv = np.meshgrid(x,y)
rad = 0.201943
design_params = np.sqrt(np.square(xv) + np.square(yv)) < rad

matgrid = mp.MaterialGrid(mp.Vector3(Nx,Ny),
                          mp.air,
                          mp.Medium(index=3.5),
                          weights=design_params,
                          do_averaging=True,
                          beta=1000,
                          eta=0.5)

geometry = [mp.Cylinder(center=mp.Vector3(0.35,0.1),
                        radius=0.1,
                        height=mp.inf,
                        material=mp.Medium(index=1.5)),
            mp.Block(center=mp.Vector3(-0.15,-0.2),
                     size=mp.Vector3(0.2,0.24,mp.inf),
                     material=mp.Medium(index=2.5)),
            mp.Block(center=mp.Vector3(-0.2,0.2),
                     size=mp.Vector3(0.4,0.4,mp.inf),
                     material=matgrid)]

sources = [mp.Source(mp.ContinuousSource(1.0),
                     center=mp.Vector3(),
                     size=mp.Vector3(0.2,0.2,0),
                     component=mp.Ez)]

sim = mp.Simulation(resolution=20,
                    cell_size=cell_size,
                    geometry=geometry,
                    sources=sources,
                    boundary_layers=[mp.PML(thickness=0.1,side=mp.High,direction=mp.Y)])

mon = sim.add_dft_fields([mp.Ez],
                         1.0,
                         0,
                         1,
                         center=mp.Vector3(0.3,-0.2),
                         size=mp.Vector3(0.3,0.1))

grid_resolution = 320
sim.plot2D(resolution=grid_resolution)
plt.title('resolution = {}'.format(grid_resolution))
plt.savefig('ring_matgrid_res{}'.format(grid_resolution),
            dpi=150,
            bbox_inches='tight')
```

**output**
```
Using MPI version 3.1, 1 processes
     cylinder, center = (0.35,0.1,0)
          radius 0.1, height 1e+20, axis (0, 0, 1)
          dielectric constant epsilon diagonal = (2.25,2.25,2.25)
     block, center = (-0.15,-0.2,0)
          size (0.2,0.24,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (6.25,6.25,6.25)
     block, center = (-0.2,0.2,0)
          size (0.4,0.4,1e+20)
          axes (1,0,0), (0,1,0), (0,0,1)
Geometric-object bounding-box tree:
     box (-0.5..0.5, -0.5..0.5, 0..0)
          bounding box (-0.4..0, 0..0.4, -5e+19..5e+19)
          shift object by (0, 0, 0)
          block, center = (-0.2,0.2,0)
               size (0.4,0.4,1e+20)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (-0.25..-0.05, -0.32..-0.08, -5e+19..5e+19)
          shift object by (0, 0, 0)
          block, center = (-0.15,-0.2,0)
               size (0.2,0.24,1e+20)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (0.25..0.45, 0..0.2, -5e+19..5e+19)
          shift object by (0, 0, 0)
          cylinder, center = (0.35,0.1,0)
               radius 0.1, height 1e+20, axis (0, 0, 1)
Geometric object tree has depth 1 and 3 object nodes (vs. 3 actual objects)

Elapsed run time = 0.5704 s
```

**Comment by smartalecH** [comment]
Is it just as fast when plotting a 2D cross section of a 3D simulation?

**Comment by oskooi** [comment]
> Is it just as fast when plotting a 2D cross section of a 3D simulation?

Taking the example above and extending it to 3d, the difference in runtimes at a resolution of 320 for this PR compared to master is 0.5773 s vs. 87.5444 s. This translates to a speedup of a factor of more than 150. Using the revamped `plot2D` in this PR to generate a 2D cross section of a 3D cell seems to be just as fast as the 2D cell shown previously (both roughly ~0.5 s).

plot2D_comparison_3d

**PR 1827**
```
Using MPI version 3.1, 1 processes
     cylinder, center = (0.35,0.1,0)
          radius 0.1, height 0.5, axis (0, 0, 1)
          dielectric constant epsilon diagonal = (2.25,2.25,2.25)
     block, center = (-0.15,-0.2,0)
          size (0.2,0.24,0.8)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (6.25,6.25,6.25)
     block, center = (-0.2,0.2,0)
          size (0.4,0.4,0.3)
          axes (1,0,0), (0,1,0), (0,0,1)
Geometric-object bounding-box tree:
     box (-0.5..0.5, -0.5..0.5, 0..0)
          bounding box (-0.4..0, 0..0.4, -0.15..0.15)
          shift object by (0, 0, 0)
          block, center = (-0.2,0.2,0)
               size (0.4,0.4,0.3)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (-0.25..-0.05, -0.32..-0.08, -0.4..0.4)
          shift object by (0, 0, 0)
          block, center = (-0.15,-0.2,0)
               size (0.2,0.24,0.8)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (0.25..0.45, 0..0.2, -0.25..0.25)
          shift object by (0, 0, 0)
          cylinder, center = (0.35,0.1,0)
               radius 0.1, height 0.5, axis (0, 0, 1)
Geometric object tree has depth 1 and 3 object nodes (vs. 3 actual objects)

Elapsed run time = 0.5773 s
```

**master**
```
Using MPI version 3.1, 1 processes
-----------
Initializing structure...
time for choose_chunkdivision = 3.6012e-05 s
Working in 3D dimensions.
Computational cell is 1 x 1 x 1 with resolution 320
     cylinder, center = (0.35,0.1,0)
          radius 0.1, height 0.5, axis (0, 0, 1)
          dielectric constant epsilon diagonal = (2.25,2.25,2.25)
     block, center = (-0.15,-0.2,0)
          size (0.2,0.24,0.8)
          axes (1,0,0), (0,1,0), (0,0,1)
          dielectric constant epsilon diagonal = (6.25,6.25,6.25)
     block, center = (-0.2,0.2,0)
          size (0.4,0.4,0.3)
          axes (1,0,0), (0,1,0), (0,0,1)
Geometric-object bounding-box tree:
     box (-0.503125..0.503125, -0.503125..0.503125, -0.503125..0.503125)
          bounding box (-0.4..0, 0..0.4, -0.15..0.15)
          shift object by (0, 0, 0)
          block, center = (-0.2,0.2,0)
               size (0.4,0.4,0.3)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (-0.25..-0.05, -0.32..-0.08, -0.4..0.4)
          shift object by (0, 0, 0)
          block, center = (-0.15,-0.2,0)
               size (0.2,0.24,0.8)
               axes (1,0,0), (0,1,0), (0,0,1)
          bounding box (0.25..0.45, 0..0.2, -0.25..0.25)
          shift object by (0, 0, 0)
          cylinder, center = (0.35,0.1,0)
               radius 0.1, height 0.5, axis (0, 0, 1)
Geometric object tree has depth 1 and 3 object nodes (vs. 3 actual objects)
subpixel-averaging is 17.152% done, 19.3209 s remaining
subpixel-averaging is 30.5805% done, 9.08038 s remaining
subpixel-averaging is 43.1867% done, 5.26365 s remaining
subpixel-averaging is 59.4996% done, 2.72321 s remaining
subpixel-averaging is 77.5109% done, 1.16075 s remaining
subpixel-averaging is 91.4617% done, 0.374134 s remaining
subpixel-averaging is 17.1149% done, 19.3729 s remaining
subpixel-averaging is 30.3951% done, 9.1621 s remaining
subpixel-averaging is 42.9373% done, 5.31743 s remaining
subpixel-averaging is 59.0481% done, 2.77456 s remaining
subpixel-averaging is 77.1773% done, 1.18323 s remaining
subpixel-averaging is 91.0708% done, 0.392198 s remaining
subpixel-averaging is 17.1217% done, 19.3657 s remaining
subpixel-averaging is 30.5299% done, 9.1024 s remaining
subpixel-averaging is 43.2002% done, 5.26053 s remaining
subpixel-averaging is 59.4356% done, 2.7301 s remaining
subpixel-averaging is 77.4469% done, 1.16553 s remaining
subpixel-averaging is 91.2022% done, 0.386007 s remaining
time for set_epsilon = 86.1741 s
-----------

Elapsed run time = 87.5444 s

```

**Comment by stevengj** [comment]
> Removed the frequency parameter from plot2D since this cannot be passed to get_epsilon_grid.

Why can't we update `get_epsilon_grid` to accept this?

**Comment by oskooi** [comment]
> Why can't we update get_epsilon_grid to accept this?

This seems like it is possible without much additional effort since `get_epsilon_grid` can just call a modified version of the recently added function [`eff_chi1inv_row_disp`](https://github.com/NanoComp/meep/blob/bc58782c0d16e6d59937028cbe2ffa402f5f24b0/src/meepgeom.cpp#L2560-L2614) (from #1801) to compute the trace of the dispersive ε tensor rather than the current function [`chi1p1`](https://github.com/NanoComp/meep/blob/bc58782c0d16e6d59937028cbe2ffa402f5f24b0/src/meepgeom.cpp#L878-L896) which computes the instantaneous ε.

**Comment by stevengj** [comment]
See how array slice gets a scalar quantity from the trace of inv(ε): https://github.com/NanoComp/meep/blob/03ea7ad3f56a5fb40b5211d69a885572e23b2f2b/src/array_slice.cpp#L379-L387

**Comment by oskooi** [comment]
Pull out and refactor:
https://github.com/NanoComp/meep/blob/03ea7ad3f56a5fb40b5211d69a885572e23b2f2b/src/meepgeom.cpp#L2568-L2588

**Comment by stevengj** [comment]
Refactor that code into new function `get_chi1_tensor_disp(meep::component c, std::complex<double> tensor[9],
  const meep::vec &r, double freq, geom_epsilon *geps)`, which is then called by `eff_chi1inv_row_disp`.  Your code will call `get_chi1_tensor_disp` directly, and then compute the trace/3 as the average "scalar" ε.

**Comment by oskooi** [comment]
Added support for dispersive materials following the approach we discussed [above](https://github.com/NanoComp/meep/pull/1827#issuecomment-971136265). This involved reinstating the `frequency` parameter of `plot2D` and adding a new `frequency` parameter to `get_epsilon_grid`. The test case below for an arbitrary collection of dispersive and non-dispersive material geometries demonstrates a speedup of more than a factor of 600 times compared to master.

get_epsilon_grid_dispersive

```py
import meep as mp
from meep.materials import SiN, Co
import numpy as np
import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as plt

mp.verbosity(3)

cell_size = mp.Vector3(1,1,1)

design_region_resolution = 500
design_shape = mp.Vector3(1.0,1.0,0)
Nx = int(design_region_resolution*design_shape.x) + 1
Ny = int(design_region_resolution*design_shape.y) + 1
x = np.linspace(-0.5*cell_size.x,0.5*cell_size.x,Nx)
y = np.linspace(-0.5*cell_size.y,0.5*cell_size.y,Ny)
xv, yv = np.meshgrid(x,y)
rad = 0.201943
w = 0.104283
design_params = np.logical_and(np.sqrt(np.square(xv) + np.square(yv)) > rad,
                               np.sqrt(np.square(xv) + np.square(yv)) < rad+w,
                               dtype=np.double)

matgrid = mp.MaterialGrid(mp.Vector3(Nx,Ny),
                          mp.air,
                          mp.Medium(index=3.5),
                          weights=design_params,
                          do_averaging=True,
                          beta=1000,
                          eta=0.5)

geometry = [mp.Cylinder(center=mp.Vector3(0.35,0.1),
                        radius=0.1,
                        height=0.5,
                        material=mp.Medium(index=1.5)),
            mp.Block(center=mp.Vector3(-0.15,-0.2),
                     size=mp.Vector3(0.2,0.24,0.8),
                     material=SiN),
            mp.Block(center=mp.Vector3(-0.2,0.2),
                     size=mp.Vector3(0.4,0.4,0.3),
                     material=matgrid),
            mp.Prism(vertices=[mp.Vector3(0.05,0.45),
                               mp.Vector3(0.32,0.22),
                               mp.Vector3(0.15,0.10)],
                     height=0.5,
                     material=Co)]

sources = [mp.Source(mp.ContinuousSource(1.0),
                     center=mp.Vector3(),
                     size=mp.Vector3(0.2,0.2,0.2),
                     component=mp.Ez)]

sim = mp.Simulation(resolution=400,
                    cell_size=cell_size,
                    geometry=geometry,
                    sources=sources,
                    boundary_layers=[mp.PML(thickness=0.1,side=mp.High,direction=mp.Y)])

mon = sim.add_dft_fields([mp.Ez],
                         1.0,
                         0,
                         1,
                         center=mp.Vector3(0.3,-0.2),
                         size=mp.Vector3(0.3,0.1))

grid_resolution = 400
frequency = 0
sim.plot2D(resolution=grid_resolution,
           frequency=frequency,
           output_plane=mp.Volume(center=mp.Vector3(),
                                  size=mp.Vector3(1,1,0)))
plt.title('frequency = {}, resolution = {}'.format(frequency,grid_resolution))
plt.savefig('ring_matgrid_3d_dispersive_res{}_freq{}_master'.format(grid_resolution,frequency),
            dpi=150,
            bbox_inches='tight')
```

File: python/visualization.py [hunk 1]
```diff
@@ -329,26 +329,30 @@ def sort_points(xy):
             return ax
     return ax
 
-def plot_eps(sim, ax, output_plane=None, eps_parameters=None):
+def plot_eps(sim, ax, output_plane=None, eps_parameters=None, frequency=None):
     # consolidate plotting parameters
     eps_parameters = default_eps_parameters if eps_parameters is None else dict(default_eps_parameters, **eps_parameters)
 
     # Determine a frequency to plot all epsilon
```
Comment (stevengj) [inline python/visualization.py:336]: I would do instead:
```py
if frequency is not None:
    warnings.warn('The frequency parameter of plot2D has been deprecated. Use the frequency key of the eps_parameters dictionary instead.')
    eps_parameters['frequency'] = frequency
if eps_parameters['frequency'] is None:
    ....
```

**Pull request merged**